src/torchmetrics/detection/iou.py

```diff
@@ -0,0 +1,168 @@
+
+        self.box_format = box_format
+        self.iou_threshold = iou_threshold
+        self.class_metrics = class_metrics
+        self.add_state("detections", default=[], dist_reduce_fx=None)
+        self.add_state("detection_scores", default=[], dist_reduce_fx=None)
+        self.add_state("detection_labels", default=[], dist_reduce_fx=None)
+        self.add_state("groundtruths", default=[], dist_reduce_fx=None)
+        self.add_state("groundtruth_labels", default=[], dist_reduce_fx=None)
+
+    def update(self, preds: List[Dict[str, Tensor]], target: List[Dict[str, Tensor]]) -> None:  # type: ignore
+        """Add detections and ground truth to the metric.
+        Args:
+            preds: A list consisting of dictionaries each containing the key-values
+            (each dictionary corresponds to a single image):
+            - ``boxes``: ``torch.FloatTensor`` of shape
+                [num_boxes, 4] containing `num_boxes` detection boxes of the format
+                specified in the contructor. By default, this method expects
+                [xmin, ymin, xmax, ymax] in absolute image coordinates.
+            - ``scores``: ``torch.FloatTensor`` of shape
+                [num_boxes] containing detection scores for the boxes.
+            - ``labels``: ``torch.IntTensor`` of shape
+                [num_boxes] containing 0-indexed detection classes for the boxes.
+            target: A list consisting of dictionaries each containing the key-values
+            (each dictionary corresponds to a single image):
+            - ``boxes``: ``torch.FloatTensor`` of shape
+                [num_boxes, 4] containing `num_boxes` ground truth boxes of the format
+                specified in the contructor. By default, this method expects
+                [xmin, ymin, xmax, ymax] in absolute image coordinates.
+            - ``labels``: ``torch.IntTensor`` of shape
+                [num_boxes] containing 1-indexed ground truth classes for the boxes.
+        Raises:
+            ValueError:
+                If ``preds`` is not of type List[Dict[str, Tensor]]
+            ValueError:
+                If ``target`` is not of type List[Dict[str, Tensor]]
+            ValueError:
+                If ``preds`` and ``target`` are not of the same length
+            ValueError:
+                If any of ``preds.boxes``, ``preds.scores``
+                and ``preds.labels`` are not of the same length
+            ValueError:
+                If any of ``target.boxes`` and ``target.labels`` are not of the same length
+            ValueError:
+                If any box is not type float and of length 4
+            ValueError:
+                If any class is not type int and of length 1
+            ValueError:
+                If any score is not type float and of length 1
+        """
+        _input_validator(preds, target)
+
+        for item in preds:
```

Comment:
in the functionals you have split it up in `compute` and `update`, but `update` only stores the inputs here. Is there any way we could already apply the update from the functionals to maybe save some memory? Otherwise there is no point of having update and compute in the functionals at all :)